lecture-notes/2-vcs.md

```diff
@@ -1764,7 +1764,205 @@ decided to give it its own subcommand, and `git restore` was born.
+   const char *right_str = argv[3];
+   errno = 0;
+   char *endptr;
+-  int left = strtol(left_str, &endptr, 10);
+-  if (left_str == endptr) {
+-    errno = EINVAL;
+-  }
+-  if (errno != 0) {
++  int left;
++  if (!to_int(left_str, left)) {
+     perror(argv[0]);
+     return EXIT_FAILURE;
+   }
+```
+
+Git sees that you've modified `main.c`, so it refuses to check out any other
+version of that file for fear of destroying your modifications! Although this is
+a nice safety feature, it can be annoying: you're not ready to commit your
+refactor---it crashes and introduces a compiler warning---but you have more
+important things to do than debug it. So what can you do?
+
+One option is to commit the unfinished change and finish it in a later commit.
+Although this violates our rule that each commit should be complete and
+self-contained, it can be a viable strategy when the refactoring work is on its
+own branch. (More on that next lecture.) Another is to destroy your work so far
+(e.g. by `git restore`ing `main.c`) and redo it later, which is clearly
+undesirable.
+
+A third option is to use a subcommand called `git stash` to save your work for
+later. `git stash` is a bit like `git commit`, in that it saves changes from the
+working tree to a more permanent location. Unlike `git commit`, however, those
+changes don't become part of the Git history. Instead, they go into a special
+area called the *stash*, which unlike the history is neither linear nor shared
+between different copies of the repository. `git stash` is shorthand for `git stash push`, which moves all uncommitted changes to a new stash entry:
+
+```console?prompt=$
+$ git stash
+Saved working directory and index state WIP on main: 0fb5235 Revert "Support the modulo operator"
+$ git status
+On branch main
+nothing to commit, working tree clean
+$ cat main.c
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+int calc(int left, char op, int right) {
+  switch (op) {
+  case '+':
+    return left + right;
+  case '-':
+    return left - right;
+  case '*':
+    return left * right;
+  case '/':
+    return left / right;
+  }
+  fprintf(stderr, "Unrecognized op `%c'.\n", op);
+  exit(EXIT_FAILURE);
+}
+
+int main(int argc, char **argv) {
+<snip>
+$ 
+```
+
+After running it, the working tree becomes clean and your changes (like the new `to_int()` function in `main.c`) disappear! Where have they gone?
+
+```console?prompt=$
+$ git stash list
+stash@{0}: WIP on main: 0fb5235 Revert "Support the modulo operator"
+```
+
+There's a new stash entry, named `stash@{0}`[^todo-footnote]. Unlike commits, stash entries store *diffs* rather than *snapshots*. When you save a stash entry, the differences between the latest commit and the working tree get saved. Then, once your other work is done and you're ready to restore that stash entry, those differences are applied to your current working tree, even if pieces of it have changed. To restore a stash entry, use `git stash pop`:
```

Comment:
same as above; as long as we don't lie outright i'm fine with pedantry in footnote

Reply:
"the UI presents" vs "Git treats/stores as"